### PR TITLE
[Beam-1705]: add method in Pipeline to visualize the DAG

### PR DESCRIPTION
**Pull request by XuMingmin**
Add a method `visualizePipeline` in `Pipeline` to visualize data flow.

### CODE REVIEW

**Comment by XuMingmin** [comment]
Hi @tgroh , can you please take a look?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java [hunk 1]
```diff
@@ -380,6 +524,32 @@ public void traverseTopologically(PipelineVisitor visitor) {
...
+   * <pre>{@code
+   * Pipeline pipeline = Pipeline.create(...);
+   * PCollection<?> stream = pipeline.apply("source",IO.<?>read())
+   *   .apply("process_step", ParDo.of(...));
+   * stream.apply("persistent", IO.<?>write());
+   * }</pre>
+   *
+   * <p>It's visualized as
+   * <pre>{@code
+   * 1. persistent
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: This seems backwards to me.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java [hunk 2]
```diff
@@ -354,6 +356,148 @@ public void visitPrimitiveTransform(TransformHierarchy.Node node) { }
...
+    class DisplayVisitor extends Defaults {
+      private List<String> pCollections = new ArrayList<>();
+      private List<String> collectionProducer = new ArrayList<>();
+      private List<TransformEdge> transformEdges = new ArrayList<>();
+
+      private List<Integer> endPoints = new ArrayList<>();
+      private boolean[][] edgeVector;
+
+      @Override
+      public void visitPrimitiveTransform(TransformHierarchy.Node transformTreeNode) {
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: I don't like ignoring all composites and only printing the flattened version of the pipeline.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java [hunk 3]
```diff
@@ -380,6 +524,32 @@ public void traverseTopologically(PipelineVisitor visitor) {
...
 
   /**
+   * retrieve {@code Pipeline}, and print out data flow in visualized format.
+   *
+   * <p>Give a pipeline like
+   * <pre>{@code
+   * Pipeline pipeline = Pipeline.create(...);
+   * PCollection<?> stream = pipeline.apply("source",IO.<?>read())
+   *   .apply("process_step", ParDo.of(...));
+   * stream.apply("persistent", IO.<?>write());
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: What does this look like if there's branches or joins? Especially if something ends with high fanout, or a large degree of independent paths?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/PipelineTest.java [hunk 4]
```diff
@@ -384,6 +384,32 @@ public boolean matches(AppliedPTransform<?, ?, ?> application) {
...
+
+    PCollection<String> left = input.apply("Left1", myTransform).apply("Left2", myTransform);
+    PCollection<String> right = input.apply("Right", myTransform);
+
+    PCollection<String> both = PCollectionList.of(left).and(right)
+        .apply(Flatten.<String>pCollections());
+
+    String visualizedFlow = pipeline.visualizePipeline();
+    Assert.assertEquals(
+        "[end] Flatten.PCollections\n"
```
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/PipelineTest.java]: I'm not super certain I'm sold on the "the complex thing displays properly", as what's likely to happen is if this test ever breaks we go in and fiddle with the expected values until it matches up with the new one (if it looks sane at all)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java [hunk 5]
```diff
@@ -354,6 +356,148 @@ public void visitPrimitiveTransform(TransformHierarchy.Node node) { }
...
+      private boolean[][] edgeVector;
+
+      @Override
+      public void visitPrimitiveTransform(TransformHierarchy.Node transformTreeNode) {
+        String operationName = transformTreeNode.getFullName();
+        List<TaggedPValue> inputTags = transformTreeNode.getInputs();
+        List<TaggedPValue> outputTags = transformTreeNode.getOutputs();
+
+        switch ((inputTags.isEmpty() ? 0 : 2) + (outputTags.isEmpty() ? 0 : 1)) {
+        case 3:// case for !input.isEmpty() && !output.isEmpty()
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: You're already writing out all of the logic, so why not use if statements.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java [hunk 6]
```diff
@@ -354,6 +356,148 @@ public void visitPrimitiveTransform(TransformHierarchy.Node node) { }
...
+          }
+        }
+      }
+
+      /**
+       * internal object to express data flow of one transform.
+       *
+       */
+      private static class TransformEdge {
+        private String inputCollections;
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: Can these not be an actual `PCollection` and convert as late as possible

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java (hunk 1 above)
Comment (XuMingmin) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: yes, it's displayed backward, retrieve back from all end PCollections

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java (hunk 3 above)
Comment (XuMingmin) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: This is an example which joins two sources, and applied two filters separately on the join result.
```
[end] filte_one/ParMultiDo(Anonymous)
    ParDo(Anonymous)/ParMultiDo(Anonymous)
        CoGroupByKey/ConstructCoGbkResultFn/ParMultiDo(ConstructCoGbkResult)
            CoGroupByKey/GroupByKey
                CoGroupByKey/Flatten.PCollections
                    CoGroupByKey/MakeUnionTable0/ParMultiDo(ConstructUnionTable)
                        Window.Into()/Window.Assign
                            ParDo(ListingItemPartExtract)/ParMultiDo(ListingItemPartExtract)
                                Values/Values/Map/ParMultiDo(Anonymous)
                                    item_topic/Remove Kafka Metadata/ParMultiDo(Anonymous)
                                        item_topic/Read(UnboundedKafkaSource)
                    CoGroupByKey/MakeUnionTable1/ParMultiDo(ConstructUnionTable)
                        Window.Into()2/Window.Assign
                            ParDo(ListingVariationPartExtract)/ParMultiDo(ListingVariationPartExtract)
                                Values2/Values/Map/ParMultiDo(Anonymous)
                                    variation_topic/Remove Kafka Metadata/ParMultiDo(Anonymous)
                                        variation_topic/Read(UnboundedKafkaSource)
[end] filter_two/ParMultiDo(Anonymous)
    ParDo(Anonymous)/ParMultiDo(Anonymous)
        CoGroupByKey/ConstructCoGbkResultFn/ParMultiDo(ConstructCoGbkResult)
            CoGroupByKey/GroupByKey
                CoGroupByKey/Flatten.PCollections
                    CoGroupByKey/MakeUnionTable0/ParMultiDo(ConstructUnionTable)
                        Window.Into()/Window.Assign
                            ParDo(ListingItemPartExtract)/ParMultiDo(ListingItemPartExtract)
                                Values/Values/Map/ParMultiDo(Anonymous)
                                    item_topic/Remove Kafka Metadata/ParMultiDo(Anonymous)
                                        item_topic/Read(UnboundedKafkaSource)
                    CoGroupByKey/MakeUnionTable1/ParMultiDo(ConstructUnionTable)
                        Window.Into()2/Window.Assign
                            ParDo(ListingVariationPartExtract)/ParMultiDo(ListingVariationPartExtract)
                                Values2/Values/Map/ParMultiDo(Anonymous)
                                    variation_topic/Remove Kafka Metadata/ParMultiDo(Anonymous)
                                        variation_topic/Read(UnboundedKafkaSource)
```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java (hunk 6 above)
Comment (XuMingmin) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: +1

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java (hunk 5 above)
Comment (XuMingmin) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: that's another option, just `switch/case` is easier to read for me.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/PipelineTest.java (hunk 4 above)
Comment (XuMingmin) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/PipelineTest.java]: also confuse to me the proper way to check the return, string comparison is odd. Any suggestion?

**Comment by XuMingmin** [comment]
Agree to slim down the API of `Pipeline`, maybe can move to a separate `PipelineUtil` class?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java (hunk 2 above)
Comment (XuMingmin) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: sorry that not quite sure 'composites'/'flattened' here, can you have more tips? are you referring to 
```java
transformTreeNode.isCompositeNode();
transformTreeNode.getEnclosingNode();
```

Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: e.g. `enter/leaveCompositeTransform`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java (hunk 5 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: I strongly prefer `ifs`

Comment (XuMingmin) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: gotcha, let me update it.

**Comment by davorbonaci** [comment]
This is a relatively complex thing to do right.

Any ideas on handling side inputs and multiple outputs?

**Comment by XuMingmin** [comment]
Have a talk with @davorbonaci , will revise the scope and implementation. 
Close it.

**Pull request closed without merging**